### ISSUE TITLE
mob done commit behaviour inconsistent, dependent on --no-squash

### ISSUE THREAD

**Issue by albran** [remotemobprogramming/mob#219]
**mob version 2.1.0**

the commit behaviour of `mob done` appears to be inconsistent depending if the `--no-squash` flag is included.

without `--no-squash`, `mob done` results in uncommitted, staged changes from the wip branch, requiring a `git commit`
with `--no-squash`, changes on the wip branch are committed to the proper branch, using the default wip commit message

this behaviour does not seem to be specified in the docs, is it intentional?
would be nice to configure `mob done` commit behaviour seperately

**Comment by gregorriegler** [comment]
> The docs specify that --no-squash simply prevents squashing of wip commits assumedly created by mob next commands. 

Ah, I see.
So you probably only want to use `mob done --no-squash` in case you have the commits lined up with correct messages in your wip branch. And you do this by using the `mob squash-wip` command.
So lets say the wip branch has the following commits:

- wip commit
- wip commit
- Refactor so and so
- wip commit
- Feature X done

You may now run `mob squash-wip` and it will get rid of all the wip commits. So you will end up with this on your wip branch:

- Refactor so and so
- Feature X done

After this, when you did `mob done` it would squash everything and remove those commits.
So in stead you do `mob done --no-squash` which will allow you to keep those commits.

I hope this clears up how `mob done --no-squash` came to be.

Now, I'm not sure about `--no-commit` and `--ff` thing. Do you have a better Idea to achieve the thing?

**Comment by albran** [comment]
in the case of our specific workflow, we simply want to keep every single commit from the mob branch when `mob done` does its work. `--no-squash` works fine for this need.

the issue i am trying to raise is the unexpected "side effect" of running `mob done --no-squash` with changes since the previous commit. in this case (as i understand it from examining the console):

1. changes are committed to the mob branch as a WIP commit, using the default WIP commit message
2. in a series of steps, the mob branch is merged into the proper branch using fast-forward (the final merge is executed by `git merge --no-commit --ff mob/my-branch` with the proper branch checked out). in doing so, the proper branch receives the WIP commit message from step 1. as its latest commit. 
3. the mob branch is deleted

from reading the docs, when i execute `mob done` with new changes since the last commit, regardless of `--no-squash`, i expect the opportunity to specify a coherent commit message. in reality, `--no-squash` deprives me of this opportunity, because of what happens in step 2.

**Comment by gregorriegler** [comment]
I think I understand now. You are loosing the possiblity to name the final commit of your mob session.
Well, I would know a way around it, by commiting it on your wip branch.
But as of now, I do not know how to achieve what you would like to do.
Perhaps it would work with stash.

**Comment by gregorriegler** [comment]
One way to achieve what you want to achieve is to rename the last commit by using --ammend. You can do that as it is not yet pushed, right?

**Comment by simonharrer** [comment]
In your wip branch, you only have commits with good commit messages. Q1: Do you use `mob next` then at all? Or how do you create the commits in your wip branch?

Before you're issuing `mob done --no-squash` you might have uncommitted changes. Q2: Would it make more sense to let the `mob done --no-squash` fail and let the user know that a manual commit is required (i.e., every change must be committed)? This might prevent having commits with the intermediate commit message in the main branch.

**Comment by albran** [comment]
hi Simon,

thank you for this fantastic tool!

our workflow goes `mob start` > `mob next` > `mob start` > `mob next` >  ... >`mob start` > `mob done --no-squash`

after a mob done, on the proper (non-WIP) branch we want to see a series of `mob next [ci-skip] [ci skip] [skip ci]` commits, except for the most recent commit, which will have a specified commit message.

to do this now we issue `mob done --no-squash -m "<specified commit message>"`. here the `-m` flag overrides the default WIP commit message, on the automatic commit that happens on the WIP branch just before it gets merged into the proper branch as part of the flow of `mob done `

we arrived at this solution after some confusion about the behaviour of `mob done`. when we execute `mob done --no-squash`, we expected the same flow as with vanilla `mob done`, that is, receive staged changes on which we can issue a `git commit`.

in the case of a vanilla `mob done,` you are squashing all WIP commits and staging everything changed since the beginning of the mob session. in the case of how we assumed `mob done --no-squash` would work, you are keeping the commit history and staging _only the changes that have happened since the last WIP commit_, ready for a `git commit`. in the case of there being no changes since the last WIP commit i guess it gets messy in trying to achieve the same state, maybe the last commit is soft reset and its changes staged, ready for a `git commit`?

i guess i am just pointing out some of the confusion we had around the behaviour of mob done and how it combines with --no-squash, because this is not really elaborated on in the docs. i also understand that we perhaps have a peculiar use case...

**Comment by gregorriegler** [comment]
I agree. I think it would be better if it worked as albran explained. Need to figure out a way to make that happen.

**Comment by gregorriegler** [comment]
So I found 2 ways to do this. One would be with the git stash, which I would rather avoid to not interfer with the users stash, so we don't surprise them.

And the other way is to just make a soft reset of the final wip commit.

I tried the latter in this pull request: https://github.com/remotemobprogramming/mob/pull/222

**Comment by gregorriegler** [comment]
@albran can you please check if this change solves your problem?
@simonharrer do you think we should go for this?

**Comment by simonharrer** [comment]
Your PR is fine for me.

**Comment by albran** [comment]
thanks @gregorriegler 

i'm trying to test this but am unfamiliar with how to install the branched version on my machine.
i cloned the repo, switched to the relevant branch and ran install.sh, though on closer inspection it looks like this file always curls the github master.
what is the correct way?

**Comment by simonharrer** [comment]
best would be to clone the repo, checkout the PR branch, and then call `./install`. This should install the version based on the currently checked out source.

**Comment by albran** [comment]
thanks @simonharrer

and thank you for the PR @gregorriegler, the behaviour is now consistent! I added a suggestion for getting the same todo prompt irrespective of squash or no squash.
